### PR TITLE
Feature/ge 1089

### PR DESCRIPTION
**Pull request by AltusBarry**
Refactor of Middleware.

Update to session data helper utils
Site is active split out to own middleware
Theme middleware sets and clears values on whitelisted urls
Session middleware sets and clears values on whitelisted urls

### CODE REVIEW

File: authentication_service/middleware.py [hunk 1]
```diff
@@ -43,55 +77,47 @@ def fetch_theme(request, key=None):
             # list.
             theme = next_query_args.get("theme", [None])[0]
 
-    if theme is None:
-        theme = get_session_data(request, key)
-
-    return theme
+    return theme.lower() if isinstance(theme, str) else theme
```
Comment (cobusc) [inline authentication_service/middleware.py]: What can `theme` be other than a string?

File: authentication_service/middleware.py [hunk 2]
```diff
@@ -16,11 +16,45 @@
...
 from authentication_service.constants import SESSION_KEYS, EXTRA_SESSION_KEY
-from authentication_service.utils import update_session_data, get_session_data
+from authentication_service.utils import (
+    update_session_data, get_session_data, delete_session_data
+)
 
 
 LOGGER = logging.getLogger(__name__)
 
+SESSION_UPDATE_URL_WHITELIST = [
```
Comment (cobusc) [inline authentication_service/middleware.py]: Maybe use a `set` for constant time `in` lookups.

File: authentication_service/middleware.py [hunk 3]
```diff
@@ -43,55 +77,47 @@ def fetch_theme(request, key=None):
...
-            except AuthorizeError as e:
-                # Suppress one of the errors oidc raises. It is not required
-                # for pages beyond login.
-                if e.error == "unsupported_response_type":
-                    pass
-                else:
-                    raise e
-
+            authorize = authorize_client(request)
+            if not isinstance(authorize, AuthorizeEndpoint):
```
Comment (cobusc) [inline authentication_service/middleware.py]: I prefer the positive check here: `if isinstance(authorize, HttpResponse):`

File: authentication_service/tests/settings/111.py [hunk 4]
```diff
@@ -82,7 +82,6 @@
     "django.middleware.locale.LocaleMiddleware",
     "django_otp.middleware.OTPMiddleware",
     "authentication_service.middleware.ErrorMiddleware",
-    "authentication_service.middleware.ThemeManagementMiddleware",
```
Comment (cobusc) [inline authentication_service/tests/settings/111.py]: Why was this removed from tests?

File: project/settings.py [hunk 5]
```diff
@@ -88,7 +88,6 @@
     "django.middleware.locale.LocaleMiddleware",
     "django_otp.middleware.OTPMiddleware",
     "authentication_service.middleware.ErrorMiddleware",
-    "authentication_service.middleware.ThemeManagementMiddleware",
```
Comment (cobusc) [inline project/settings.py]: Why is this removed?

Comment (cobusc) [reply project/settings.py]: Oh, I see it moved to base.

File: authentication_service/tests/settings/111.py (hunk 4 above)
Comment (cobusc) [reply authentication_service/tests/settings/111.py]: Oh, I see it moved to base.

**Comment by cobusc** [comment]
Left some comments

File: authentication_service/middleware.py (hunk 1 above)
Comment (AltusBarry) [reply authentication_service/middleware.py]: Should not be anything other than a string. Will explicitly cast to string.

File: authentication_service/middleware.py [hunk 6]
```diff
@@ -104,69 +130,73 @@ def create_disabled_site_response(request):
...
                 )
 
+
+class SessionDataManagementMiddleware(MiddlewareMixin):
+    """
+    NOTE: This Middleware should always be as near the end of the Middleware
+    list in settings. Middleware is evaluated in order and this needs to happen
+    as near the end as possible.
+    """
+    client_uri_key = SESSION_KEYS["redirect_client_uri"]
```
[the remark this answers is not in the document]
Comment (AltusBarry) [reply authentication_service/middleware.py]: I like that. Thanks.

[the remark this answers is not in the document]
Comment (AltusBarry) [reply authentication_service/middleware.py]: These keys aren't specifically for this middleware class or middleware in general. They are just keys that need to stay constant, hence `constants.py`.

The values can potentially be set else where too. This middleware just happens to be the only place that sets them.

Making use of a constant dictionary is a habit from having had to make use of shared keys between many different Django apps. If you'd prefer an object vs a dictionary, that can be arranged. However I will keep it in `constants.py`

[the remark this answers is not in the document]
Comment (cobusc) [reply authentication_service/middleware.py]: Yes, the crux was that it is not necessary to look up the name of something you want using a hardcoded string value. Then you might as well hardcode the looked up value. So it is the use of a dictionary that is not nice IMHO.
I'm very happy that these are defined in `constants.py`

File: authentication_service/constants.py [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
 EXTRA_SESSION_KEY = "ge_session_extra_data"
 
-SESSION_KEYS = {
-    "redirect_client_uri": "ge_oidc_client_uri",
-    "redirect_client_name": "ge_oidc_client_name",
-    "redirect_client_terms": "ge_oidc_client_terms",
-    "theme": "theme"
-}
+
+class SessionKeys:
```
Comment (cobusc) [inline authentication_service/constants.py]: Thanks

File: authentication_service/views.py [hunk 8]
```diff
@@ -40,7 +40,7 @@
 from authentication_service.user_migration.models import TemporaryMigrationUserStore
 
 
-REDIRECT_SESSION_KEY = constants.SESSION_KEYS["redirect_client_uri"]
+REDIRECT_SESSION_KEY = constants.SessionKeys.CLIENT_URI
```
Comment (cobusc) [inline authentication_service/views.py]: Please change this variable name to be something more appropriate (e.g. `CLIENt_URI`)

**Comment by cobusc** [comment]
:+1: when last nitpick is sorted out.
